allennlp/tango/dataloader.py

```diff
@@ -0,0 +1,186 @@
+        return self.num_batches_per_epoch()
+
+
+class DataLoaderAdapter(DataLoader):
+    """Adapts a TangoDataLoader to an old-school AllenNLP DataLoader."""
+
+    def __init__(self, tango_data_loader: TangoDataLoader):
+        self.tango_data_loader = tango_data_loader
+        self.target_device: Optional[torch.device] = None
+
+    def __len__(self) -> int:
+        result = self.tango_data_loader.num_batches_per_epoch()
+        if result is None:
+            raise ValueError(
+                "DataLoaderAdapter cannot be used with TangoDataLoaders "
+                "that don't have a defined epoch length. Old-school AllenNLP "
+                "DataLoaders don't support this feature, so we can't adapt to it."
+            )
+        return result
+
+    def __iter__(self) -> Iterator[TensorDict]:
+        if self.target_device is None:
+            return iter(self.tango_data_loader)
+        else:
+            for batch in iter(self.tango_data_loader):
+                yield move_to_device(batch, self.target_device)
+
+    def iter_instances(self) -> Iterator[Instance]:
+        raise NotImplementedError()
+
+    def index_with(self, vocab: Vocabulary) -> None:
+        raise NotImplementedError(
+            "In AllenNLP Tango, your instances have to be indexed before they get to the data loader."
+        )
+
+    def set_target_device(self, device: torch.device) -> None:
+        self.target_device = device
+
+
+@TangoDataLoader.register("batch_size")
+class BatchSizeDataLoader(TangoDataLoader):
+    """A data loader that turns instances into batches with a constant number of instances
+    per batch.
+
+    * `instances` contains the instances we want to make batches out of.
+    * `batch_size` is the number of instances per batch
+    * `drop_last` specifies whether to keep the last batch in case it is smaller than
+      `batch_size
+    * `shuffle` specifies whether to shuffle the instances before making batches"""
+
+    def __init__(
+        self,
+        instances: Sequence[Instance],
+        batch_size: int,
+        drop_last: bool = False,
+        shuffle: bool = True,
+    ):
+        self.instances = instances
+        self.batch_size = batch_size
+        self.drop_last = drop_last
+        self.shuffle = shuffle
+
+    def num_batches_per_epoch(self) -> Optional[int]:
+        batch_count = len(self.instances) / self.batch_size
+        if self.drop_last:
+            return floor(batch_count)
+        else:
+            return ceil(batch_count)
+
+    def __iter__(self) -> Iterator[TensorDict]:
+        instances: Sequence[Instance]
+        if self.shuffle:
+            instances_as_list_just_to_make_mypy_happy = list(
+                self.instances
+            )  # make a new list pointing to the same instance objects
+            random.shuffle(instances_as_list_just_to_make_mypy_happy)
+            instances = instances_as_list_just_to_make_mypy_happy
+        else:
+            instances = self.instances
```

Comment:
It might be really expensive to create a list here if your actual dataset is backed by something like an Arrow table. Alternatively, you could just shuffle a list of indices, and then iterate over the list of indices.

Reply:
Introducing `ShuffledSequence` 71411e15b9ab206960bffe4aeac624ec953bc31c